### ISSUE TITLE
Provide a way to load resources without running scripts

### ISSUE THREAD

**Issue by dploeger** [godotengine/godot-proposals#4925]
### Describe the project you are working on

I'm the core developer of [EgoVenture](https://egoventure.deep-entertainment.de/), a first person adventure game framework and a contributor to [Escoria](https://github.com/godot-escoria), a third person adventure game framework. Both projects are using Resources to provide a savegame functionality.

### Describe the problem or limitation you are having in your project

As recommended in the [official Godot docs](https://docs.godotengine.org/en/stable/tutorials/scripting/resources.html#creating-your-own-resources), Resources should be used to store and retrieved structured data.

This can be used e.g. to persist the current game's state (aka savegame).

As many godot developers ([GDquest, among others](https://twitter.com/NathanGDQuest/status/1549641252467089408?s=20&t=NxIUyG6D-ARFW_PG3i25aQ)) has pointed out: Using resources to store data can be dangerous as resources can also hold scripts which are executed when loading the resource later.

That way, malicious people can send seemingly valid savegames and embed malicious code.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

As it can be required to load scripts from Resources, Godot should feature a `SecureResourceLoader`, which specifically skips loading/executing scripts from resources so developers can still rely on resources as the most optimised method for storing structured data.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Very simply put, `SecureResourceLoader` should extend `ResourceLoader`, override the `load`/`load_interactive` methods and not load and execute any scripts in the given Resource.

(this might sound way easier than it is to actually embed it, I still have to dig through the code)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, because loading resources is a core functionality.

### Is there a reason why this should be core and not an add-on in the asset library?

Loading resources is a core functionality and security should be a core responsibility as well.

**Comment by dploeger** [comment]
Hm, so I guess one way to implement it would be by adding a `load_secure` method to `ResourceFormatLoader` and e.g. clean the scripts from resource definitions when loading them e.g. in [ResourceFormatText](https://github.com/godotengine/godot/blob/8adf04804550957f43c810b13fcce66dd76e5fd8/scene/resources/resource_format_text.cpp#L390). The `SecureResourceLoader` would call `load_secure` instead of `load`.

Alternatively, we could implement this into the `load` function of `ResourceLoader` by adding a boolean secure parameter which defaults to false (to be backwards compatible)...

**Comment by YuriSizov** [comment]
We've been discussing it a bit, and it should load scripts, because otherwise you lose the benefits of scripted resources. It should ignore bundled scripts and external scripts loading from the outside of `res://`.

Another option would be to validate the loading resource against some sort of schema. I can imagine you can refer to your internal type as a schema and the engine would strictly validate serialized members against it. This way you can avoid extra fields being added and extra types being used, preventing the possibility of a scriptless attack.

**Comment by dploeger** [comment]
The schema would not work **that** great in my context because it's variable for each game designed with the framework so adding a schema would put that burden on every developer using the framework. 

Maybe instead something like an `allow_scripts` field in the Resource class which could be set to false or something?

**Comment by YuriSizov** [comment]
> The schema would not work that great in my context because it's variable for each game designed with the framework so adding a schema would put that burden on every developer using the framework.

A schema would be resource based. So ideally it would require zero work from the developer beyond designing their resources.


> Maybe instead something like an `allow_scripts` field in the Resource class which could be set to false or something?

Scripts are not the only attack vector here. And it shouldn't be a resource decision, it should be a loader decision. Because otherwise the bad actor can still replace the script in your resource that allows it with their script. Which is why it should filter out scripts based on their quality while loading: if they are bundled or if they are external and not from `res://`.

**Comment by dploeger** [comment]
@YuriSizov good call

**Comment by TheDuriel** [comment]
Just to bring this over from twitter:

It is my opinion that the correct way to solve this, would be to completely disallow embedding script within any kind of godot resource format. Be that resource, scene, or others. Yes, this would include removing the entire embedded script functionality. (Which already lost most of its usefulness due to lack of feature support for things like class names.)

(I been theorizing a way to inject code through the icon.png and default audio bus files present in all projects. Completely outside of the control of the developer.)

**Comment by mieldepoche** [comment]
> this would include removing the entire embedded script functionality 

I personnally use them sometimes, at least when prototyping.

> I been theorizing a way to inject code through the icon.png and default audio bus files present in all projects. Completely outside of the control of the developer.

Do you mean, corrupting a pck file? You can litterally inject anything you want inside it really easily.

**Comment by TheDuriel** [comment]
No, this wouldn't touch the PCK at all. But it's irrelevant in any case.

**Comment by NathanLovato** [comment]
I'd also lean towards removing embedded scripts, which could be part of solving this, for all their limitations. The way they come to bite you when your project grows and you get errors in embedded scripts, for instance, you quickly lose the little time they saved you initially.

But that likely depends on finding some consensus, I remember quite a few people liked them.

**Comment by Calinou** [comment]
> I'd also lean towards removing embedded scripts, which could be part of solving this, for all their limitations. The way they come to bite you when your project grows and you get errors in embedded scripts, for instance, you quickly lose the little time they saved you initially.

This was proposed in the past and rejected: https://github.com/godotengine/godot/pull/54553

Since then, built-in scripts have been improved significantly, so I wouldn't remove them (even though I don't use them myself).

**Comment by Sslaxx** [comment]
Is this also potentially an "abuse" of `Resource` and its ilk anyway, i.e. it was never intended to be used in this fashion and so any such use will be unsupported? If so, might be worth rewording the documentation to discourage this practice.

**Comment by NathanLovato** [comment]
I think it doesn't matter that it wasn't _initially_ meant to be used that way if it's possible to support the use case. Resources are a great tool for modding, bringing in user-created data into Godot in general, they have the potential to save you a lot of code and coupling (because of how there's this built-in load() system that'll load resources only once), they can work as text in debug builds for quick bug fixing, and binary in release builds for smaller file sizes and faster loading...

They have pros, they're just lacking several small improvements, in this case for user safety.

**Comment by NathanLovato** [comment]
> This was proposed in the past and rejected: https://github.com/godotengine/godot/pull/54553
> 
> Since then, built-in scripts have been improved significantly, so I wouldn't remove them (even though I don't use them myself).

Makes total sense, yeah, if the community uses them then built-in scripts are here to stay.

**Comment by SilencedPerson** [comment]
> I'd also lean towards removing embedded scripts, which could be part of solving this, for all their limitations. The way they come to bite you when your project grows and you get errors in embedded scripts, for instance, you quickly lose the little time they saved you initially.
> 
> But that likely depends on finding some consensus, I remember quite a few people liked them.

Maybe it could be useful to not allow embedded scripts on load?

**Comment by NathanLovato** [comment]
Yes that's already an idea in this conversation, see these two comments by @YuriSizov:

- https://github.com/godotengine/godot-proposals/issues/4925#issuecomment-1190110784
- https://github.com/godotengine/godot-proposals/issues/4925#issuecomment-1190162080

**Comment by dalexeev** [comment]
Personally, I don't see a problem with the `.tres`, `.res`, etc. formats supports script embedding. The problem only arises when this format is used for purposes for which it was not intended.

In my opinion, `.tres` is not intended for save files and similar external files, the purpose of which is to store data, and not extend the functionality of the game. And it's not just security, but other aspects that I can talk about, but they are beyond the scope of this discussion.

And as mentioned above, you **can** safely use `Resource` for external files, but you need to use `ResourceFormatLoader` for that, not `.tres`/`.res`.

At the same time, I am completely in favor of having an analog of the `allow_objects` parameter in `str2var`, `ConfigFile`, etc.

**Comment by Sslaxx** [comment]
Custom resources (in 3.x) require C++ modules, which may be beyond the scope of some people's intended usage, or skills.

**Comment by YuriSizov** [comment]
> In my opinion, `.tres` is not intended for save files and similar external files, the purpose of which is to store data, and not extend the functionality of the game.

That's pretty arbitrary. Resources are just scriptable objects with built-in serialization capacity. In C# land you can serialize any class with XML in the same way, and it's used in games all the time. I think, that's the go-to way in Unity, at least from my experience with digging into game saves of games made with it.

**Comment by dploeger** [comment]
That's also exactly what the docs say. 😉

**Comment by dploeger** [comment]
> Custom resources (in 3.x) require C++ modules, which may be beyond the scope of some people's intended usage, or skills.

No, you can just extend the Resource class and you get a custom resource. I heavily make use of that in EgoVenture.

**Comment by Sslaxx** [comment]
This may be somewhat addressed by https://github.com/godotengine/godot-proposals/issues/5010

**Comment by Proggle** [comment]
One of Godot's big strengths is streamlining common gamedev tasks, so you don't have to spend time reinventing the wheel. Saving and loading the game is one of the most common functions that devs need to implement (up there with having collision detection), but currently it isn't as straightforward as it could be.

And of course, from the game developer perspective, as far as the player's experience goes, the only real question that needs to be answered is "What data goes into the savefile?".  Whatever method the developer uses to cram the data into a save file, and then pull it back out unchanged, will function equally well.

So, since this is a common feature where the exact algorithm doesn't matter to the end product, the developer should be able to just trust a built-in Godot feature to do it properly, and not worry about what's going on under the hood.  

As is, resources are almost the perfect way for Godot to handle saving and restoring states. It uses Godot's unique strengths to handle savegames easily, and doesn't require futzing around with special cases to handle unsupported datatypes (like Vector2s in json).  

It's unfortunate that this method is currently unsafe, but if inputs can be sanitized, it's a perfect way to never need to think about the underlying savegame/loadgame algorithms ever again.

**Comment by TheDuriel** [comment]
Or just, save the resource using File, not the ResourceSaver, and encrypt it.

Or, since your savegame boils down to just being a dictionary anyways. (And lets be honest, custom Resources are just fancy typed dictionaries.) Just use File.store_var()

Resources as savegames only blipped up recently as a consideration, since I think people have finally caught on that json text files are a poor idea. But the next step is raw binary data dumps. Saving and loading a resource is actually more (or at least identical) work than making a dictionary and saving/loading that.

But that isn't even the topic of this issue. The topic is the glaring security problem caused by Godot automatically loading resources that could contain malicious code. Which is only a problem in two cases: A developer using resources where normal files would be sufficient. Or a malicious agent placing resources where it will override the pck contents to inject code. (Which requires the developer to have omitted the res.// in the file path.)

**Comment by dalexeev** [comment]
I agree that `File.store_var/get_var` in combination with a dictionary is already sufficient for a convenient implementation of save files. Although the analog of `allow_objects` should be added to the `VariantParser`, so that text serialization (for example, `ConfigFile`) also becomes safe.

`tscn`/`tres` are not suitable for save files, not only because of security, but also because of compatibility. This format is simply not intended for such purposes, as it contains references to external resources and contains nested resources, that is, it makes the save file dependent on their implementation. Attempting to load a save file based on resources from an old incompatible version of the game will result in an error loading the resource, while json, `ConfigFile` or a serialized dictionary can be loaded without error and then converted to the new version.

**Comment by lesleyrs** [comment]
It should be no problem to use resources in this way if you only export to html right? Since nobody will try to use others potentially harmful save files on web?

**Comment by alekh-ai** [comment]
Is there any way to use unique tokens for custom resources. That way it will be safer to use.

**Comment by Calinou** [comment]
> Is there any way to use unique tokens for custom resources. That way it will be safer to use.

Are you referring to encryption or signing of custom resources? That's a topic for a different proposal. Godot already offers way to save/load encrypted files, but I doubt there's a built-in signing mechanism you could make use of.

**Comment by reptofrog** [comment]
It would've been great to be able to granulally restrict what loaded scripts can and what they can't do, something that was proposed in here: #4642 in conjunction with #5010

**Comment by derkork** [comment]
I have created a small [library ](https://godotengine.org/asset-library/asset/2249) that should solve this problem for saved games. It basically works similar to the whitelist @lyuma  posted, but it will flat-out refuse to load any resource file that contains resources of type GDScript. This will probably not be good enough for user generated content where you may want to allow some sandboxed scripting, but it should work decently for saved games which usually will not contain any inline scripts, so it's a nice 80/20 solution I think.

**Comment by TheDuriel** [comment]
A resource is entirely capable of pointing towards a script definition outside of itself, or even outside of the project.

**Comment by derkork** [comment]
Yes it is. But for this to work, the player would have needed to download and place this script in addition to the save game file on their computer, so I wouldn't think this is a practical attack. That being said, if you have comments regarding the library it may be better to [open up an issue](https://github.com/derkork/godot-safe-resource-loader/issues) there instead of  using this discussion thread.

**Comment by lyuma** [comment]
I'm sorry I forgot to link this here.
Last month I wrote a GDScript resource parser which can be used for validation before loading the resource, without any need for engine support:
https://gist.github.com/lyuma/8de4620a402d565b86e1287150c8fb31

It's a bit scary in that it's security critical code, but on the plus side, in this form it would be fairly easy to audit and ensure it functions identically to the resource loader.

The risk of this approach is if Godot's resource parser ever diverges from this code (one if statement is different, for example), then that difference in code path could be used to create an exploit. My expectation is Godot would increase FORMAT_VERSION if it changes the format, but if its parsing code is not careful and they add the new features to the old format, it would create an exploit.

I think this approach would be safe if it explicitly checks for a specific Godot major and minor version (e.g. 4.2) and human review is done on each version to ensure it continues to match parsing logic. Personally, I'm willing to take this approach because I think it is less error prone than the failure case of accidentally using an unpatched engine, or having validation built into Godot and having security critical code decay by mistake during an unrelated change.

Let me know what you think about this design.

**Comment by theraot** [comment]
Since nobody seems to have mentioned it, I'll point out that if loading resources can skip scripts it would also address https://github.com/godotengine/godot/issues/65393

**Comment by Mohamed-Kr** [comment]
I don't have a very deep understanding of what can be done or not, 
but maybe we can change the Resource class, we could give it functions like filter_scripts_on_load()->bool and smth like allowed_scripts_on_load()->Array, that would give information on what can be loaded or not and is called only if the former function returns true

And we could have ResourceLoader.safe_load that would take something like a class as an argument and load only if what is loaded is of that class, and then follow the behaviour for the loading of the scripts provided by the loaded Resource

**Comment by addmix** [comment]
PackedScenes separate loading from instantiation, and I believe a similar approach should be possible with Resources as a whole. I suggest that the ResourceLoader can have an optional parameter to load resources (and by extension, PackedScenes), but while prohibiting/delaying script initialization/execution.

It's not quite the same, but I think PackedScene's GenEditState design concept could be used for this purpose. If you can load/instantiate items, and modify them before any script is allowed to run, you could filter out any scripts or unwanted content, before any execution has occurred.

The usage would be like this:

`ResourceLoader.load("user://path/to/UGC.res", "", CacheMode.CACHE_MODE_REUSE, LoadingMode.LOAD_MODE_EDIT)`

the `LoadingMode.LOAD_MODE_EDIT` would cause scripts to be prevented from all forms of execution, until the resource is allowed to continue script execution.

This also naturally opens the door for scripting to filter/process resources and scenes before they can effect anything within the system.

**Comment by gabrielesilinic** [comment]
I'd like to pop in for a minute to give my suggestion. 

probably stripping down scripts could be kind of a bad idea for some use cases. 

The optimal way would have been having gdscript work more like javascript and having a sandboxed language for the most part. unfortunately this seems to be impossible. So the laziest but most effective solution would be to sign every single script that is part of the project (or have scripts be a resource that does not get directly packed into the save but just linked).

this can be as simple as a list of hashes separated by the script (unique) name on even on a text file possibly ordered just for the sake of binary search. 

This way you would not lose any benefit coming from resources but you would prevent arbitrary code execution (since only a set amount of very specific scripts can be ran).

also by having the thing be a list any mod manager script could add their own scripts to such list. obviously this relies on the user's intelligence that he won't run any malicious script which would add itself to such signature registry. but at this point if signing is compromised in such a way it becomes the user's problem, because he ran something untrusted where the attack vector could have been a thousand other things.

though if we know the user is extra dumb we may want to have a few options where to put maybe an encryption key (like the registry in windows) just so everything breaks whenever a user arbitrarily replaces the signature registry file beside the game without putting much thought into the process, it's fine if a willing user or script willingly ran by the user can retrieve the key from the registry or whatever other place itself.

This is not 100% foolproof but should prevent the least determined fools from getting in trouble.

**Comment by Pidermonkey24** [comment]
I'm not entirely sure if this is even remotely possible but would it be possible to have a parameter that would not allow loaded resources to use certain libraries like OS, HTTPS, or Filesystem? This definitely wouldn't be fool hardy in the slightest and I'm unsure if it's even possible in 4.x but it would likely deter a non zero amount of bad actors.

**Comment by Calinou** [comment]
> I'm not entirely sure if this is even remotely possible but would it be possible to have a parameter that would not allow loaded resources to use certain libraries like OS, HTTPS, or Filesystem?

Yes, but this is not the topic of this proposal (which is only about not running scripts *at all*). See https://github.com/godotengine/godot-proposals/issues/5010 instead.

**Comment by jkulawik** [comment]
Perhaps worth pointing out, or adding to the "Saving games" tutorial, is that Godot's JSON offers these very useful methods:
- `stringify` - can dump Dictionary as string, but complex types (e.g. Transforms) cannot be parsed back AFAIK
- `from_native` / `to_native` - allows conversion of all (?) the core types from Godot

I ran a little test and was not able to save a script like this, but don't take it as definitive proof that loading one is not possible.

Combining these methods seems like a decent alternative for Resources if you're selective about what you save; I was able to convert my Resource saves to JSON and get it to mostly[^1] work remarkably easily[^2]. The only data where I needed to use the native conversion so far was Transforms.

This method still needs a bunch of work to convert any complex data you save (e.g. turn Dictionaries to custom resources) so I would still much rather be able to dump and read Resources instead of dealing with JSON.

[^1]: Some of my complex data types don't load yet and I will have to asses them individually

[^2]: With the caveat that I already put in a bunch of work much earlier that made it easy; I made it a point to save specific dictionaries instead of just dumping objects when I wanted to reduce the number of external resources being saved in my `.tres` file. Meaning, I was already not relying on Resources too heavily in the first place.

**Comment by typed-sigterm** [comment]
# SECURITY ALERT

For those who need to safely load `.tres` files, please note that **one `[ext_resource]` can contain multiple `path`s**, as shown in GHSA-5vf5-j43p-g226. If you're using [godot-safe-resource-loader](https://github.com/derkork/godot-safe-resource-loader), please update to v0.1.1 ASAP.

**Comment by h0lley** [comment]
addressing the argument of using FileAccess instead:

take the scenario where it is **not** desirable for the savegame to be binary or encrypted.

there's valid cases where you want a savegame to be as readable as possible, such as when you deliberately want to offer your players an easy way to modify it.

imho, TRES has much nicer readability than JSON (and of course better type support), so using ResourceSaver for savegames is desirable here.

in terms of user-side implementation, being able to just dump a Resource as a savegame and not having to bother with converting between dictionaries would be quite nice, too, albeit just a bonus.

I wonder what is blocking the implementation here? looking at derkork's and typed-sigterm's godot-safe-resource-loader, it seems like at least one possible solution is pretty straight forward. it is clearly a feature broadly desired, too, sitting on the second page of most liked proposals.

**Comment by TheDuriel** [comment]
> imho, TRES has much nicer readability than JSON (and of course better type support), so using ResourceSaver for savegames is desirable here.

.tres serialization includes adding thousands of escape characters, and turning binary resources into huge byte arrays.

**Comment by Cretezy** [comment]
Hey! I've made a library which can serialize and deserialize almost anything (any custom script/resource and Godot built-ins) to both JSON and binary, safely. Here's the link and a quick example:

https://github.com/Cretezy/godot-object-serializer

```gdscript
class Data:
	var name: String
	var position: Vector2


func _init() -> void:
	# Required: Register possible object scripts
	ObjectSerializer.register_script("Data", Data)

	# Setup data
	var data := Data.new()
	data.name = "hello world"
	data.position = Vector2(1, 2)

	var json = DictionarySerializer.serialize_json(data)
	""" Output:
	{
		"._type": "Object_Data",
		"name": "hello world",
		"position": {
			"._type": "Vector2",
			"._": [1.0, 2.0]
		}
	}
	"""

	data = DictionarySerializer.deserialize_json(json)
```

**Comment by Shadowblitz16** [comment]
This is why real oop should be done instead of scripts.

If gdscript classes compiled down to actual types and we couldn't add scripts to objects this would be alot less of an issue if one at all.

Then resources would just be pure data.

**Comment by abrossimow** [comment]
I see this issue coming up often in the community - in the context of saving and loading, as well as modding. The points brought up here are very insightful.

In my opinion the old, [proposed approach](https://github.com/godotengine/godot-proposals/issues/4925#issuecomment-1190083917) of a separate `load_secure` still holds up. Resources (and other data) loaded this way would not be allowed to carry embedded or attached scripts. This would offer developers the possibility to allow script loading from data that actually needs it (i.e. mods or UGC) and restrict injection from other sources (i.e. save files) on a more selective basis.

**Comment by TheDuriel** [comment]
> I see this issue coming up often in the community - in the context of saving and loading, as well as modding. The points brought up here are very insightful.

Please use officially endorsed methods of handling saving instead. Resource based savegames have been propagated by ill informed content creators only, and some even issued retractions in the past.

https://docs.godotengine.org/en/stable/tutorials/io/saving_games.html

This resource issue does not need to exist for your project. It is a self imposed problem.

**Comment by h0lley** [comment]
If it weren't for the security concern of embedded scripts, then resources would be the superior format for savegames and without a doubt in many or perhaps most applications, should then be the recommended, standard choice.

The README.md of [Godot Safe Resource Loader](https://github.com/derkork/godot-safe-resource-loader) summarizes it quite nicely:

> Why not just use JSON or another format instead of resources?
> Resources provide excellent support for storing/loading large graphs of nested objects and keeping references between these objects intact. This is very useful for savegames, as you can simply build a nice object graph in your game and then save / load it with a single line of code. If you want to do the same thing with JSON you will need to manually serialize/deserialize all objects into JSON structures and also find a way of keeping references within the object graph intact. This is a lot of work, just to replicate something that resources already provide out of the box. I therefore think, that using this library is an acceptable compromise between security and implementation effort.

Ultimately, we'll likely have to wait for the concept of GDScript as Resource to be [dropped altogether](https://github.com/godotengine/godot-proposals/issues/14652).

Until then, for people happy with TRES files, the aforementioned addon does the trick.

**Comment by TheDuriel** [comment]
> If it weren't for the security concern of embedded scripts, then resources would be the superior format for savegames and without a doubt in many or perhaps most applications, should then be the recommended, standard choice.

There are no advantages conferred by using Resources for save games. It's more code and worse for versioning. The only upgrade to store_var() would be structs if they ever get merged.

.tres files are exclusively editor facing.

**Comment by dalexeev** [comment]
> Resources provide excellent support for storing/loading large graphs of nested objects and keeping references between these objects intact.

Note that, in my opinion, it poses problems not only from a security standpoint, but also with compatibility/versioning of your save files. You're tightly coupling the current structure of your save files with the internal organization of your project.

If you need to change the architecture of your game in the future, or even just rename some variables, move files, etc., your old save files will be incompatible with new versions of your game. Unlike JSON, you can't first load the raw data and then parse it, converting the old structure to a new one. The resource loader will attempt to deserialize the TRES and load the required subresources, which may no longer exist at the old paths.

Resources can also serialize data that isn't essential for saving. This can lead to a violation of the integrity of the saved state, which may occur when deserializing a resource in a new version of the game (for example, if you export some interdependent properties and in the new version the nature of this dependency has changed or the algorithm for calculating the dependent property has changed). Even if this doesn't cause any bugs, it can lead to an unreasonable increase in the size of the save file.

**Comment by addmix** [comment]
> > Resources provide excellent support for storing/loading large graphs of nested objects and keeping references between these objects intact.
> 
> Note that, in my opinion, it poses problems not only from a security standpoint, but also with compatibility/versioning of your save files. You're tightly coupling the current structure of your save files with the internal organization of your project.
> 
> If you need to change the architecture of your game in the future, or even just rename some variables, move files, etc., your old save files will be incompatible with new versions of your game. Unlike JSON, you can't first load the raw data and then parse it, converting the old structure to a new one. The resource loader will attempt to deserialize the TRES and load the required subresources, which may no longer exist at the old paths.
> 
> Resources can also serialize data that isn't essential for saving. This can lead to a violation of the integrity of the saved state, which may occur when deserializing a resource in a new version of the game (for example, if you export some interdependent properties and in the new version the nature of this dependency has changed or the algorithm for calculating the dependent property has changed). Even if this doesn't cause any bugs, it can lead to an unreasonable increase in the size of the save file.

I absolutely agree with those limitations of Resources. But this proposal somewhat suggests that Resources *should* have the ability to be loaded as raw data (like a dictionary or JSON) without code execution, initialization, or modification.

Though unrelated to loading Resources without code execution, there are still ways to version and update Node and Resource class configurations in a way similar to how you would version/update/reinitialize a JSON format. This is something you may need to do even in an editor context with Nodes/Resources in the inspector, not only for save games in shipped project. Here's an example which I implemented previously: https://github.com/addmix/godot_aerodynamic_physics/blob/0abdbea3e7f35cd1d6dd6809c741fe491df85913/core/aero_influencer_3d/aero_influencer_3d.gd#L403

As a summary, you provide an array of property update configurations (property string name, property type, and a function which provide the logic to update the given property to the current version), when Godot calls `Object._get_property_list()`, these property update configs are added to the property list, to "spoof" those properties into existence, so that when godot finally calls `Object._set()` for all properties saved in the Resource, those now removed/changed properties can be updated to the current version.

Though keep in mind that the entire discussion about whether JSON or Resources are better/worse, more/less work is unrelated to this proposal.

**Comment by ParadoxV5** [comment]
> If you need to change the architecture of your game in the future, or even just rename some variables, move files, etc., your old save files will be incompatible with new versions of your game. Unlike JSON, you can't first load the raw data and then parse it, converting the old structure to a new one. The resource loader will attempt to deserialize the TRES and load the required subresources, which may no longer exist at the old paths.

Mods that patch game files *because the game doesn’t have a JSON interface* will also break.

**Comment by addmix** [comment]
> > If you need to change the architecture of your game in the future, or even just rename some variables, move files, etc., your old save files will be incompatible with new versions of your game. Unlike JSON, you can't first load the raw data and then parse it, converting the old structure to a new one. The resource loader will attempt to deserialize the TRES and load the required subresources, which may no longer exist at the old paths.
> 
> Mods that patch game files _because the game doesn’t have a JSON interface_ will also break.

This is irrelevant to the proposal at hand. The discussion of Resource vs JSON is pedantic, and not productive for this proposal.

**Comment by TheDuriel** [comment]
The existence of better alternatives is fundamental to any discussion about resources being misused.

This proposal and any discussion about resources like this should be shut down*, because it's a case of using a hammer to drive a screw. And not only does the engine offer a perfectly fine screwdriver, it offers a deluxe edition too.

Do we need to support every possible, not actually ever acknowledged or endorsed, use case people come up with for an engine feature?

Text based resources are intended for the editor. There are use cases in which they may be loaded 'externally', like from a secondary PCK in the case of a DLC or mod package.

*Proposals are of course, never shut down. They are left to languish unread for years until someone digs them back up.

---

For an 'actual solution', mind you that it is perfectly possible to use file encryption with get and store_var() in FileAccess, **and to save resources using this method.** So if you really want to use resources as a bad version of a struct, you can, and it can be safe. And it takes practically no effort to do so.

It truly is trivial to use the tools the engine offers for, every single use case, people have come up with where they keep trying to use resources instead. https://theduriel.github.io/Godot/Saving-and-Loading#2-storing-the-data

**Comment by addmix** [comment]
It's just not a productive discussion here. Just leave a thumbs down reaction if you disagree and move on. Places like the discord and rocket chat are intended for discussions like this. It seems you are very opinionated, and that's ok, I won't try to change your mind.

Though I'm going to start with: from that blog post you shared about save data via JSON, given the limited context of the snippets shared, your code using JSON most likely still contains ACE vulnerabilities, considering that `JSON.parse_string()` is capable of loading any Variant, including Objects, of which Scripts, Resources, and Nodes derive. I'd suggest you do some testing, to ensure you aren't shipping the same vulnerability that you seek to prevent

Using Resources safely is easy: https://godotengine.org/asset-library/asset/2249

Interestingly enough, this safe resource loader approach can also work with `.tscn` files. Which for my project, leads to the most effective method for saving games being as simple as:
```gdscript
var save_game():
	var saved_scene := PackedScene.new()
	var packing_error = saved_scene.pack(get_tree().current_scene)
	if packing_error: pass #some error checking here
	var save_error = ResourceSaver.save(saved_scene, "user://game_save.tscn")
	if save_error: pass #some error checking here
```
This covers serialization for about ~95% of my entire project, with the remaining 5% being Autoload/singletons, which don't serialize by default (which may be necessary, though I try to avoid stateful singletons), and some odd cases here and there that may need some custom serialization logic.

> Text based resources are intended for the editor.

I'm not sure where you got this idea from, is demonstrably untrue, you can load `.res` and `.tres` files from any directory that you have read access to, even in exported projects. Resources are simply godot's generic data format. I would argue, so long as you are saving resources to disk, with the intent to load them in the future, you are using Resources correctly.

> For an 'actual solution', mind you that it is perfectly possible to use file encryption with get and store_var() in FileAccess

Considering that the decryption key would necessarily have to be included in the exported project binary, this is not a valid security strategy. It may prevent low-level tampering, but regardless, Godot games are trivial to inject modloaders into, and to decompile.

If you'd like, I think it would be more productive to have a discussion like this in a discord channel. I am in the official Godot discord, under the same username, so feel free to reach out if you have more questions.

**Comment by TheDuriel** [comment]
I concur, there's nothing productive to be gained here. You didn't even read the article you're criticizing. In which, I do not recommend using JSON and indeed do the opposite.

You appear dead set on using a hammer to drive screws. There's no helping you, and any security issues you encounter along the way are of your own making.

**Comment by dalexeev** [comment]
> This is irrelevant to the proposal at hand. The discussion of Resource vs JSON is pedantic, and not productive for this proposal.

This isn't quite a correct statement. You're trying to use the TRES/RES format, which was designed for completely different purposes (storing resources within a project, including loading and embedding scripts) and lacks support for the feature you require (safely loading resources from the user's file system).

And then you ask for this feature to be added to the loader, ignoring the goals and design of this format, as well as the fact that the tasks you require can be accomplished more efficiently using other means. This (exaggerated) is like complaining that a car can't fly and asking for a flight feature, ignoring the fact that a car is designed for other purposes and that a plane or helicopter should be used for flight.

Also, I'm not saying that resources themselves (the `Resource` class and its descendants, plus `ResourceSaver` and `ResourceLoader`) are bad for save files. I'm only saying that the TRES/RES format is not designed for storing/loading save files. Potentially, you could write your own resource serializer and deserializer (`ResourceFormatSaver` and `ResourceFormatLoader`) that would save/load resources into a custom save file format, eliminating script embedding and, optionally, taking into account the considerations I mentioned above.

From a usage perspective, the difference is minor (`load("user://save.mysave")` instead of `load(user://save.tres")`). From an implementation perspective, the difference is that you would have full control over the saving/loading process and be able to map the project's internal organizational structure to some abstract save file data structure. And even if the initial structure changed, you could account for it in the saver/loader, thus ensuring that the save file from the older version of the game would remain compatible with the new version.

However, I personally see no advantage in using `ResourceFormatSaver` and `ResourceFormatLoader` over directly using `FileAccess`. Both approaches require roughly the same amount of code.

**Comment by addmix** [comment]
First, I would just like to say, I'm not irrationally against custom game-save methods. I'm sure there is plenty of positive reason to use them, but as a plugin developer, game-saves are only one of multiple reasons why I would like to see this proposal implemented (see my Resource versioning/updating implementation, mentioned above). This is my primary reasoning for saying that the fixation on using Resources for game saves is irrelevant, and honestly, pretty pedantic. The responses providing workarounds, while some are helpful, miss the point because this proposal is beneficial to any usage of Resources, both in editor and in exported projects. I have not read any downsides or limitations as to why this proposal couldn't/shouldn't be implemented.

> You're trying to use the TRES/RES format, which was designed for completely different purposes (storing resources within a project, including loading and embedding scripts)

Says who? Resources are the de facto method of serialization within Godot. I see absolutely no reason why an arbitrary line should be drawn at game saves. They are just another type of data, and Resources, if nothing more, help you save and load them.

The main thing that these save-game serialization alternatives seem to miss is that custom serialization requires an order of magnitude more work than Resources do, which only makes sense, due to how closely Resources are integrated into Godot.  In the context of testing your project in the editor, running a scene is indistinguishable from running your game from the main scene and loading into a specific level, with a specific game state. The logic and implementation of a game save is already there, it's inherent to the behavior of your project.

Resources require:
- Implementation in game to trigger saving/loading
- - Relatively few exceptions that may need extra serialization logic (autoload singletons, non-Resource Object types (which probably shouldn't be used in a serialization context anyways)

The rest of the work is already done:
- - No up front cost to implement, considering serialization/deserialization logic is done in the existing setter/getter code of `@export`ed properties, which is already implemented assuming your game functions properly during editor testing.
- - ResourceSaver omits default values from serialization (reduces save file size)
- - Loading/deserialization uses the same logic/implementation as regular scene loading, matching in-editor testing (better testing, better parity between running in-editor and in exported projects)

A custom serialization approach requires (at minimum):
- Implementation in game to trigger saving/loading
- Serializer/deserializer implementation (non-trivial, if you want features to minimize game save sizes, will be significantly slower than ResourceLoader/Saver unless written in C#/C++)
- Each serialized object in the project requires serialize/deserialize implementation, which would be boilerplate, or mostly duplicate of what is already implemented via `@exports` and setters/getters (this may include lists of property names that must be serialized, boilerplate functions to apply saved values to the object)
- - Deserialization code happens after scene loading, after _init(), _enter_tree(), and _ready() (deserialization code may require extra re-initialization code to ensure the correct state, increasing complexity a tremendous amount)
- - Deserialization code may require extra loading/instantiation steps for in-game objects/entities that may be added or removed to the game during runtime. (now it becomes a responsibility problem, dynamic entities must be tied to specific spawner nodes, which must keep track of all remaining entities, and be responsible for serialization. Which becomes more complicated as, on deserialization, the dynamic entities must be initialized to the exact nodepath, with the correct name, which must be done reliably before the dynamic entities can be serialized, leading to a race condition)

I'm sure there are a few shortcuts that can be taken to simplify and work around some of these downsides, but really, is there a tangible benefit to the added effort? What about the additional effort to debug and exhaustively test the system? It seems like in reality, even the best non-Resource save systems still lack some of the features of Resources, while still requiring a substantial amount of serialization-specific code to function. Again, why not leverage the powerful, convenient, and efficient serialization that already exists?

**Comment by enderprism** [comment]
> your code using JSON most likely still contains ACE vulnerabilities, considering that JSON.parse_string() is capable of loading any Variant, including Objects, of which Scripts, Resources, and Nodes derive

`JSON.parse_string()` only parses valid JSON, and not Objects, which can't be represented in JSON. The return type of `parse_string()` is Variant because GDScript doesn't have union types (it would be a union like `Dictionary | Array | float | bool | String | Nil`), see: https://github.com/godotengine/godot/blob/6d6e822c689acb54bf86e19cc2bbbe3fa567b123/core/io/json.cpp#L422-L464

Other than that, I agree that there should be natively be a way to parse resources without loading scripts. I don't like the idea that a script can be bundled in a file supposed to represent data, but I see the use of built-in scripts for PackedScenes.

**Comment by Naros** [comment]
> Resources are the de facto method of serialization within Godot. I see absolutely no reason why an arbitrary line should be drawn at game saves. They are just another type of data, and Resources, if nothing more, help you save and load them.

You're tightly associating `Resource` as `.tres` or `.res`, and that's the subtle differentiation that @dalexeev was trying to convey. If you want to use resources, the engine provides all the tools necessary for you to do it using resources, but in a way that satisfies all your own requirements: custom `ResourceFormatSaver` and `ResourceFormatLoader`.

What people don't often think about is using the Resource subsystem this way comes with a host of side effects that you need to be prepared to handle or understand how to handle so that Godot properly works with the `Resource`:

* Accounting for the `ResourceCache`
* Multiple open/close calls on the resource type to read UIDs, updating rename references, etc.
* And all the other noise that the resource format loader/saver comes with

So yes, Resources can provide you with a lot of nifty features OOTB, but they also come with these side effects that you need to be prepared to handle.

For game save files in particular, just using `FileAccess` directly avoids all that noise.

**Comment by AThousandShips** [comment]
I think it's important to be aware that resource management and save management are largely different topics with different goals and constraints, and a lot of how the resource system in Godot works is built around needs, goals, and constraints that aren't relevant for save files, which makes using it at minimum inefficient

Using resource files outside of `res://` also has a number of limitations and weaknesses that are important to keep in mind, which together with the already mentioned things means it's generally better to not use that system for save file purposes

**Comment by h0lley** [comment]
I want to store a (optionally) human-readable user-facing file in a format that keeps at least built-in type information, ideally custom resource types as structs as well.

to my knowledge, that's not achievable with the FileAccess helpers.

the assumption that in every kind of project, it is a suitable and feasible option to flatten and restore everything to and from dicts seems pretty off to me.

granted, I have not looked much into ResourceFormatSaver/Loader, although I feel like this shouldn't be necessary as everything already seems to be in place in vanilla Resource.

**Comment by TheDuriel** [comment]
> I want to store a (optionally) human-readable user-facing file in a format that keeps at least built-in type information, ideally custom resource types as structs as well.
> 
> to my knowledge, that's not achievable with the FileAccess helpers.
> 
> the assumption that in every kind of project, it is a suitable and feasible option to flatten and restore everything to and from dicts seems pretty off to me.
> 
> granted, I have not looked much into ResourceFormatSaver/Loader, although I feel like this shouldn't be necessary as everything already seems to be in place in vanilla Resource.

ConfigFile does this.

**Comment by h0lley** [comment]
I guess ConfigFile fits my specification, but offers no structuring beyond sectors.

I may look at `JSON.to/from_native` instead, although if that should technically satisfy my needs, it will definitely remain perplexing to me as to why when I need to serialize structures of Godot built-in typed properties, I somehow cannot use Godot's own format.

**Comment by dalexeev** [comment]
> I have not looked much into ResourceFormatSaver/Loader, although I feel like this shouldn't be necessary as everything already seems to be in place in vanilla Resource.

Please don't confuse resources as objects, their serialization formats, and file extensions. The same resource type can be saved in several different formats, each associated with its own file extension. Sometimes a single extension can correspond to multiple formats, but if I recall correctly, this only applies to imported resources (`ResourceImporter`), not loaded resources (`ResourceFormatLoader`).

Godot has two universal resource formats: TRES and RES, with corresponding file extensions. These formats support subresources, including embedded scripts. Some resource types support additional formats with their associated extensions. And for any resource type, you can add custom formats with their own file extensions. Custom formats may or may not support subresource serialization (and other resource features, like name and UID), depending on the authors' intent.

Now, let's imagine we've reached a consensus that this feature is necessary. How exactly should it work? The author suggests adding a whole new singleton, `SecureResourceLoader`, which doesn't seem good. Alternatively, we could add a flag/parameter to the `ResourceLoader.load()` method or a new `load_secure()` method. But in either case, the resource loader has no idea which resources are secure and which are not. These are properties of the formats themselves, not their loading options. We could add a virtual method to format savers/loaders, `_is_support_secure_load()`, but it's still not a good idea to rely on this information.

So, in my opinion, the most appropriate potential solution would be to introduce new formats, TRESSec and RESSec, with corresponding file extensions (similar to the HTTP and HTTPs protocols). This wouldn't require complete code duplication; the format saver/loader would just need to take the file extension into account.

However, I question why Godot should support this out of the box? If the only difference is that you use `load("user://save.tres_sec")` instead of `load("user://save.tres")`, why can't you use `load("user://save.mysave")`? Godot already provides the ability to implement custom formats. If you don't want to do it yourself, someone else can create and publish such an addon, and usability won't suffer. It doesn't have to be part of the engine, and the save format doesn't have to match an existing one.

Finally, providing this functionality out of the box means the engine developers are taking responsibility for ensuring that this format is truly secure. We don't want to give users a false sense of security. And, this is much more difficult than it seems, due to the peculiarities of C++ (buffer overflow attacks, etc.).

For reference: I previously tried to add a similar parameter for text serialization and created a [PR](https://github.com/godotengine/godot/pull/80585). But reduz (one of the engine's founders and technical lead) explained to me that logical checks aren't enough. To ensure true security, we need to minimize the attack surface, keeping all critical code as isolated as possible so that it can be fully audited at any time. This works for binary serialization, but not for text serialization, including the TRES format. My PR is still unmerged, and thanks to reduz's explanation, I understand why.

**Comment by KoBeWi** [comment]
> the assumption that in every kind of project, it is a suitable and feasible option to flatten and restore everything to and from dicts seems pretty off to me.

I do happen to hane an addon for that: https://github.com/KoBeWi/Godot-Ref-Serializer
It turns objects into Dictionaries ready for serializing. It supports nested objects and internal classes. You can use them as pseudo-structs.

> ConfigFile does this.

ConfigFile is also vulnerable, and not well suited for nested structures. As of now, the only "safe" to use text format is native JSON, which is extremely verbose.

> My PR is still unmerged, and thanks to reduz's explanation, I understand why.

And I don't xd
Why does `native_to_json()` have `with_objects` parameter, but `var_to_str()` can't?

btw, custom formats can also be vulnerable if you don't design them properly. It's not a perfect solution either.

**Comment by ParadoxV5** [comment]
> game-saves are only one of multiple reasons why I would like to see this proposal implemented
> ⸺ https://github.com/godotengine/godot-proposals/issues/4925#issuecomment-4321469394

Another use case is an interface to player-created levels, characters, etc., where both the base game and these plugins can leverage the same pure-Godot SDK and neither has to fiddle with JSON.

But, beware,
> > Mods that patch game files _because the game doesn’t have a JSON interface_ will also break.
> 
> This is irrelevant to the proposal at hand.
> 
> ⸺ https://github.com/godotengine/godot-proposals/issues/4925#issuecomment-4320682150

---

P.S. Plugins that want to inject scripts will require #5010, assuming that proposal’s not less viable.

**Comment by addmix** [comment]
> btw, custom formats can also be vulnerable if you don't design them properly. It's not a perfect solution either.

Your entire message goes pretty well with my line of thinking.

Similarly, I think it's beyond the scope of the regular resource loader to secure vulnerabilities in custom formats.

With the talk of other formats, how many other formats in Godot have this same vulnerability? I can only think of scenes with the option for built-in scripts. Most other resource formats in Godot have a fairly small scope that I can't imagine would/should allow them to contain and parse GDScript.

**Comment by dalexeev** [comment]
> > My PR is still unmerged, and thanks to reduz's explanation, I understand why.
> 
> And I don't xd Why does `native_to_json()` have `with_objects` parameter, but `var_to_str()` can't?

That's a good question. Maybe the difference is that `JSON.to_native()` does not interact with untrusted data initially, it's called after the gate has already been passed:

```gdscript
func decode_data(string, allow_objects = false):
    return JSON.to_native(JSON.parse_string(string), allow_objects)
```

That is, a vulnerability could be in `JSON.to_native()`, `JSON.parse_string()` or anywhere in the engine. While binary serialization is critical code (designed for multiplayer, etc.). `JSON.to_native()` cannot be an input interface in principle, while text serialization potentially can. But I agree that this is too complex an explanation for users. We should probably add a comment about binary serialization being better hardened from a security standpoint.

**Comment by ettiSurreal** [comment]
I think a hidden issue is that the `Saving Games` documentation page is not very beginner friendly and may be very intimidating for newcomers. It describes a process of what I assume is making a Source style quick save system, while I imagine someone who's just starting out may just want to know how to store a high score or what levels the player completed.
So I think the guide should either be simplified or split in two, with the first part being just how to save and load a dictionary on disk with `store/get_var` (since that's also the simplest way to do it), with an addendum on how to make it human-readable using JSON.

I'm bringing this up because situations like this can very easily prompt users into finding and using alternative solutions, not realizing the disadvantages or even dangers of.
With that out of the way, I still don't think Resources should be made a viable solution because fundamentally they are not designed for this. Their *main appeal* is that they can store code.

Let's think of what a Resource is. An Object is (very simplified) a type of Variant that can store arbitrary code in it, a RefCounted extends Object to add automatic memory management, and finally a Resource extends RefCounted to add the ability to save it as a file in your project or embedded into a scene (you cannot do `@export var Object` or `@export var RefCounted`)
The way you'd use a resource is to, for example, make a node that generates a mesh based on code, but want the ability to easily hot swap the code used to generate the mesh instead of putting everything in a single script. So what you do is make a resource to be used as a base, make an export variable for that resource in your node, and inherit from that resource and write a mesh generation script for every type of mesh you want to generate. Congratulations, you just roughly reinvented MeshInstance3D and the Mesh resource.
Unfortunately someone can also write a script that destroys your computer when the resource is initialized, thus any file that contains code that the engine can execute without user input.
(And because resources can be embedded into scenes this makes full scenes unsafe to save as well by extension)

So what is my point? If you want to ensure safety you have to make sure any files that may potentially be exchanged cannot be used to execute arbitrary code without user input (otherwise I think what you wanna look for is sandboxing), so what you want is to only store simple data, and we already have ways to do it. Resources are, again fundamentally not designed for this purpose and trying to force them into being viable will just leave you with the worst option of what you could already do, while time could be better spent figuring out if there's a way of making the existing options more convinent to use, or adding support for more existing formats.

**Comment by Zireael07** [comment]
>  how to store a high score or what levels the player completed.

For that, a ConfigFile is enough

**Comment by clarence112** [comment]
> For that, a ConfigFile is enough

ConfigFile suffers from the exact same vulnerability.

**Comment by addmix** [comment]
> Unfortunately someone can also write a script that destroys your computer when the resource is initialized, thus any file that contains code that the engine can execute without user input.

Some good points are made here. But it makes me realize, what is the actual use-case for scripts embedded into a Resource file? I don't mean external script subresources. You can embed an entire script within the Resource. This is the logic used for built-in scripts in scenes, a very handy feature, but I don't see a legitimate usage otherwise that would warrant this being exposed in the generic Resource format. Disabling embedded scripts in this context could fix the ACE vulnerability when loading Resources.

The threat signature for ACE resource files is easy because it contains this embedded code, which isn't normal for any Resource type that is implemented in your game. A resource file will reference the UID of the script within the project. Resources being able to reference scripts is fundamentally different than actually storing code via "built-in scripts"

**Comment by gabrielesilinic** [comment]
btw this is my opinion and just an opinion but I believe Godot could benefit from having a built-in quite solid save system as good as what the creation engine had (it is just impressive). it is just very hard to get that as high performance and as consistent without engine support I believe.

**Comment by h0lley** [comment]
> With that out of the way, I still don't think Resources should be made a viable solution because fundamentally they are not designed for this.

> and finally a Resource extends RefCounted to add the ability to save it as a file in your project or embedded into a scene (you cannot do `@export` var Object or `@export` var RefCounted)

they way Resource is presented in the documentation is simply as a data container, and its main USP is trivial writing and reading via ResourceSaver and ResourceLoader.

but ultimately, I don't really care if it's called Resource or something else, I just think that there should be a way to use Godot's own format to serialize structured data without boilerplate. structured as in it's not all flattened to nested, untyped Dictionaries and can be easily restored to an instance of a custom class, whether that's based on Resource or just RefCounted.

**Comment by bmorledge-hampton19** [comment]
As someone who is relatively new to Godot, I can confirm that what drove me to use a resource-based save system was the lack of another option that provided the same degree of convenience and flexibility. At the time, I didn't understand the intention behind resources or the limitations of using them as a save system. I saw the basic options, like FileAccess or JSON parsing, and thought to myself: "Ok, cool. Here are some low-level interfaces if I need to design things from the ground up. Now where's the solution that the Godot community has designed for the average use-case?"

When I discovered the ResourceSaver singleton, I thought for sure I'd found that solution. Create a script that extends Resource, mark all the "savable" properties with `@export`, save with ResourceSaver, load with the load function. Easy! Right? It wasn't until much later, after stumbling upon discussions like this one, that I realized the resource system wasn't intended to be used that way...

I agree that Godot would benefit from a dedicated save system that is convenient, flexible, and secure. Yes, JSON parsing works. Yes FileAccess works. However, the reality is that some users are being drawn to resource-based save systems for a reason, so perhaps the proper fix is to provide a comparable system that is actually designed for managing save data.

That being said, I also feel like we've strayed off topic for this issue. Is there a Godot proposal somewhere that describes such a save system? If not, is it perhaps time to create one?

**Comment by KoBeWi** [comment]
There is #14140 and #7041

**Comment by hsandt** [comment]
Related: **Add sandboxing for "untrusted" resource packs**
https://github.com/godotengine/godot-proposals/issues/389

**Comment by poyribn** [comment]
In my own Games, i will implement an INI / TOML / ConfigFile inspired Loader, instead of a JSON Loader,
because notating even slightly complex Datatypes (like Vector2) is tedious in JSON.

**Comment by rinart73** [comment]
Godot could really use some sort of built-in schema solution that would have serialization (including custom objects), validation (correct property types) and "parsing" (turning dictionaries back into object instances for example).
This would allow to both have safe to load save-files on a disk AND easily validate rpc arguments.
